File: frontend/src/types/savings.ts
export enum SavingsType {
  Savings = 1,
  Stock = 2,
  Gold = 3,
  Crypto = 4,
  RealEstate = 5,
  Fund = 6,
  Other = 7,
}

export enum SavingsStatus {
  Active = 1,
  Matured = 2,
  Closed = 3,
}

export const SAVINGS_TYPE_LABELS: Record<SavingsType, string> = {
  [SavingsType.Savings]:     "Tiết kiệm",
  [SavingsType.Stock]:       "Chứng khoán",
  [SavingsType.Gold]:        "Vàng",
  [SavingsType.Crypto]:      "Crypto",
  [SavingsType.RealEstate]:  "Bất động sản",
  [SavingsType.Fund]:        "Quỹ đầu tư",
  [SavingsType.Other]:       "Khác",
};

export const SAVINGS_STATUS_LABELS: Record<SavingsStatus, string> = {
  [SavingsStatus.Active]:  "Đang hoạt động",
  [SavingsStatus.Matured]: "Đã đáo hạn",
  [SavingsStatus.Closed]:  "Đã đóng",
};

export interface SavingsAccount {
  id: string;
  name: string;
  type: SavingsType;
  typeName: string;
  status: SavingsStatus;
  statusName: string;
  principalAmount: number;
  currentValue: number;
  profitLoss: number;
  interestRate?: number;
  startDate: string;
  maturityDate?: string;
  note?: string;
  createdAt: string;
}

export interface CreateSavingsAccountDto {
  name: string;
  type: SavingsType;
  principalAmount: number;
  interestRate?: number;
  startDate: string;
  maturityDate?: string;
  note?: string;
}

export interface UpdateSavingsAccountDto {
  name: string;
  type: SavingsType;
  interestRate?: number;
  maturityDate?: string;
  note?: string;
}

export interface SavingsHistoryDto {
  id: string;
  savingsAccountId: string;
  transactionType: string;
  amount: number;
  currentValueAfter: number;
  profitLossAfter: number;
  note?: string;
  date: string;
}

export interface SavingsDepositDto {
  amount: number;
  date: string;
  note?: string;
}

export interface SavingsWithdrawDto {
  amount: number;
  date: string;
  note?: string;
}

export interface UpdateSavingsValueDto {
  newValue: number;
  date: string;
  note?: string;
}

export interface SavingsInterestDto {
  amount: number;
  date: string;
  note?: string;
}

export interface CloseAccountDto {
  note?: string;
}

export interface SavingsSummary {
  totalPrincipal: number;
  totalCurrentValue: number;
  totalProfitLoss: number;
  profitLossPercentage: number;
  activeCount: number;
  totalCount: number;
}

File: ios/ExpenseTrackerApp/Shared/Models/Savings.swift
import Foundation

// MARK: - Enums

enum SavingsType: Int, Codable, CaseIterable {
    case savings = 1
    case stock = 2
    case gold = 3
    case crypto = 4
    case realEstate = 5
    case fund = 6
    case other = 7

    var displayName: String {
        switch self {
        case .savings: return "Savings"
        case .stock: return "Stocks"
        case .gold: return "Gold"
        case .crypto: return "Crypto"
        case .realEstate: return "Real Estate"
        case .fund: return "Investment Fund"
        case .other: return "Other"
        }
    }

    var sfSymbol: String {
        switch self {
        case .savings: return "banknote.fill"
        case .stock: return "chart.line.uptrend.xyaxis"
        case .gold: return "star.fill"
        case .crypto: return "bitcoinsign.circle.fill"
        case .realEstate: return "house.fill"
        case .fund: return "chart.pie.fill"
        case .other: return "creditcard.fill"
        }
    }
}

enum SavingsStatus: Int, Codable, CaseIterable {
    case active = 1
    case matured = 2
    case closed = 3

    var displayName: String {
        switch self {
        case .active: return "Active"
        case .matured: return "Matured"
        case .closed: return "Closed"
        }
    }
}

// MARK: - Response Models

struct SavingsAccount: Codable, Identifiable {
    let id: String
    let name: String
    let type: SavingsType
    let typeName: String
    let status: SavingsStatus
    let statusName: String
    let totalDeposited: Double
    let currentValue: Double
    let profitLoss: Double
    let interestRate: Double?
    let startDate: String
    let maturityDate: String?
    let note: String?
    let createdAt: String

    var isProfit: Bool { profitLoss >= 0 }
}

struct SavingsHistory: Codable, Identifiable {
    let id: String
    let savingsAccountId: String
    let transactionType: Int     // Gap: int values undocumented; use transactionTypeName for display
    let transactionTypeName: String
    let amount: Double
    let previousValue: Double
    let newValue: Double
    let profitLoss: Double
    let note: String?
    let date: String
    let createdAt: String?
}

struct SavingsSummary: Codable {
    let totalDeposited: Double
    let totalCurrentValue: Double
    let totalProfitLoss: Double
    let totalProfitPercent: Double
    let activeCount: Int
    let maturedCount: Int
    let closedCount: Int
}

// MARK: - Request Models

struct CreateSavingsAccountRequest: Codable {
    let name: String
    let type: Int
    let initialAmount: Double
    let interestRate: Double?
    let startDate: String
    let maturityDate: String?
    let note: String?
}

struct UpdateSavingsAccountRequest: Codable {
    let name: String
    let type: Int
    let interestRate: Double?
    let maturityDate: String?
    let note: String?
}

struct SavingsDepositRequest: Codable {
    let amount: Double
    let date: String
    let note: String?
}

struct SavingsWithdrawRequest: Codable {
    let amount: Double
    let date: String
    let note: String?
}

struct UpdateSavingsValueRequest: Codable {
    let newValue: Double
    let date: String
    let note: String?
}

struct SavingsInterestRequest: Codable {
    let amount: Double
    let date: String
    let note: String?
}

struct CloseAccountRequest: Codable {
    let note: String?
}
